src/net/api.js

```diff
@@ -0,0 +1,151 @@
+const http = require("http");
+const util = require("util");
+const {URL} = require("url");
+const UserApi = require("../user/userApi");
+const Database = require("../database/database");
+
+class Api {
+	constructor() {
+		this.stack = [];
+		this.apis = [];
+	}
+
+	/**
+	 * Registers a new path and matching method to the api. This path will then be checked when handling an incoming 
+	 * request. if the request matches this path and method it will call the method with the request, response and path 
+	 * data as arguments. The path data contains information about the url of the incoming request. The path may add
+	 * arguments using ":NAME", which wille be included in the path data object. For example: the request 
+	 * "/user/123/rename" will be matched to the path "/user/:id/rename" and the path data will be: {id: 123}. When an
+	 * asterisk is added to the end of the specified path the handler will match all requests that begin with the text
+	 * before the asterisk. By default the method matches all incoming methods, specified as "*". The callback may 
+	 * return a boolean, true when the stack may continue and false if the request is finished.
+	 * @param {String} path The path to register.
+	 * @param {String} method The method to register.
+	 * @param {Function} callback The callback to register.
+	 */
+	register(path, method = "*", callback) {
+		this.stack.push({
+			path, method, callback
+		});
+	}
+
+	/**
+	 * Unregisters the specified path.
+	 * @param {String} path The path to unregister.
+	 */
+	unregister(path) {
+		const handlerIndex = this.stack.findIndex(handler => handler.path === path);
+
+		if(handlerIndex < 0) return;
+
+		this.stack.splice(handlerIndex);
+	}
+
+	/**
+	 * Initializes the api by creating the server, starting error handing and creating all children api's.
+	 */
+	initialize({database} = {}) {
+		this.server = http.createServer((req, res) => {
+			this.handle(req, res).catch(err => this.handleError(err));
+		});
+		this.server.on("error", err => this.handleError(err));
+		this.apis.user = new UserApi();
+		this.apis.user.initialize(this);
+	}
+
+	/**
+	 * Handles the incoming request. This function may only be used internally as it is called by the server when there 
+	 * is a new incoming request.
+	 * @param {http.IncomingMessage} req The incoming request.
+	 * @param {http.ServerResponse} res The outgoing response.
+	 * @returns {Promise} A promise that resolves when the request and response have been handled.
+	 */
+	async handle(req, res) {
+		const url = new URL(req.url, "http://localhost/");
+
+		for(let handler of this.stack) {
+			if(handler.method !== "*" && handler.method !== req.method) continue;
+
+			const pathData = this.parsePath(handler.path, url.pathname);
+
+			if(pathData === undefined) continue;
+
+			const canContinue = await handler.callback(req, res, pathData);
+
+			if(!canContinue) return;
```

Comment:
Not sure why it's returning here, did you mean continue?

```suggestion
			if(!canContinue) continue;
```

Reply:
canContinue might be a wrong choice of words. The idea is that if a callback returns true it may continue looking for another handler in the stack. This enabled middleware to maybe check for authentication, etc.

Reply:
If it returns, the `res.end()` is never called. Is it suppose to be called by the function returning `true`?